pyccl/halos/pk_4pt.py

```diff
@@ -477,166 +425,144 @@ def halomod_Tk3D_SSC(cosmo, hmc,
+            if (prof, prof2, prof12_2pt) == (prof3, prof4, prof34_2pt):
+                dpk34[ia] -= dpk12[ia]
             else:
-                i02_34 = hmc.I_0_2(cosmo, k_use, aa, prof3, prof34_2pt, prof4)
-            P_34 = norm34 * (pk * i11_3 * i11_4 + i02_34)
-
-            if prof3 == prof1:
-                b3 = b1
-            elif prof3.is_number_counts:
-                b3 = i11_3 * norm3
-
-            if prof4 == prof2:
-                b4 = b2
-            elif prof4.is_number_counts:
-                b4 = i11_4 * norm4
-
-            dpk34[ia, :] -= (b3 + b4) * P_34
-
-    if use_log:
-        if np.any(dpk12 <= 0) or np.any(dpk34 <= 0):
-            warnings.warn(
-                "Some values were not positive. "
-                "Will not interpolate in log-space.",
-                category=CCLWarning)
-            use_log = False
-        else:
-            dpk12 = np.log(dpk12)
-            dpk34 = np.log(dpk34)
+                dpk34[ia] -= counterterm(prof3, prof4, prof34_2pt,
+                                         norm3, norm4, i11_3, i11_4)
 
-    tk3d = Tk3D(a_arr=a_arr, lk_arr=lk_arr,
+    dpk12, dpk34, use_log = _logged_output(dpk12, dpk34, log=use_log)
+
+    return Tk3D(a_arr=a_arr, lk_arr=lk_arr,
                 pk1_arr=dpk12, pk2_arr=dpk34,
                 extrap_order_lok=extrap_order_lok,
                 extrap_order_hik=extrap_order_hik, is_logt=use_log)
-    return tk3d
+
+
+def _allocate_profiles(prof, prof2, prof3, prof4, prof12_2pt, prof34_2pt,
+                       normprof1, normprof2, normprof3, normprof4):
+    """Helper that controls how the undefined profiles are allocated."""
+    if prof2 is None:
+        prof2 = prof
+    if prof3 is None:
+        prof3 = prof
+    if prof4 is None:
+        prof4 = prof2
+    if prof12_2pt is None:
+        prof12_2pt = Profile2pt()
+    if prof34_2pt is None:
+        prof34_2pt = prof12_2pt
+
+    # TODO: Remove for CCLv3.
+    if normprof1 is not None:
+        with UnlockInstance(prof):
+            prof.normprof = normprof1
+    if normprof2 is not None:
+        with UnlockInstance(prof2):
+            prof2.normprof = normprof2
+    if normprof3 is not None:
+        with UnlockInstance(prof3):
+            prof3.normprof = normprof3
+    if normprof4 is not None:
+        with UnlockInstance(prof4):
+            prof4.normprof = normprof4
+
+    return prof, prof2, prof3, prof4, prof12_2pt, prof34_2pt
+
+
+def _get_counterterm(pA, pB, p2pt, nA, nB, i11_A, i11_B, *,
+                     cosmo, hmc, k, a, pk):
+    """Helper to compute counter-terms."""
+    # p : profiles | p2pt : 2-point | n : norms | i11 : I_1_1 integral
+    bA = i11_A * nA if isinstance(pA, ProfNC) else np.zeros_like(k)
+    bB = i11_B * nB if isinstance(pB, ProfNC) else np.zeros_like(k)
+    i02 = hmc.I_0_2(cosmo, k, a, pA, prof2=pB, prof_2pt=p2pt)
+    P = nA * nB * (pk * i11_A * i11_B + i02)
+    return (bA + bB) * P
```

Comment:
I was going to use it in unit testing (I replaced the tkkssc unit test completely and the previous one had a similar implementation) but I didn't use it after all and coverage didn't change. 

I don't want to define it inside of the function because that's a waste of resources, but let me get back to you in this too. I'll probably just remove the function and put the code in the.

Reply:
I ended up just bringing it inside of the function.